### DISCUSSION TITLE
Calling a method what are the right parameters ?

### DISCUSSION THREAD

**Discussion by JaSicX** [eclipse-milo/milo#882] [General]
Hi,

we use Milo 0.6.3 for calling OPC-UA methods. (this is a followup to the previous #879 discussion and thanks for the workaround it worked)
But now we have a new problem I want to call a method but the last argument is what I think a node (its like an struct) and now I have the problem that I dont know how to compose the arguments. To correctly call the method, it always throws an exception "status=Bad_InvalidArgument, message=One or more arguments are invalid." 

The method that is the problem is the OPCUA_methodTaskBring_IDB.Method and its 4 argument (COORD). The values that have to be handed over are all integers (L, S, X, Y, Z). 
![UAexpert](https://user-images.githubusercontent.com/43895907/133415637-8adf01df-b4aa-460b-850d-712e7470aea4.PNG)

The code for calling the Method is as follows. 

```    java
public Variant[] callMethode(String methodName, String node, Variant[] inputValues) throws UaRuntimeException, UaException {
                
        Variant[] output = null;
        AddressSpace addressSpace = client.getAddressSpace();
        UaObjectNode objectNode = addressSpace.getObjectNode(NodeId.parse(node));
        UaMethodNode methodNode = (UaMethodNode) addressSpace.getNode(NodeId.parse(methodName));        
        
        // overwrite inputparameter
       inputValues = new Variant[] {new Variant((short) 1), new Variant((short) 1),
                    new Variant(true), new 
 Variant("ns=3;s=\"OPCUA_methodTaskBring_IDB\".\"UAMethod_InParameters\".\"COORD\"")};
       


        UaMethod method;
        try {
            method = methodNode.readInputArgumentsAsync().thenCombine(
                    methodNode.readOutputArgumentsAsync(),
                    (in, out) -> new UaMethod(client, objectNode, methodNode, in, out)
                ).get();

            output = method.call(inputValues);
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        } 

        return output;

    }
```

So the question for me now what I have to hand over at the 4 parameter I currently thought I have to hand over the nodeID but this doesn't work. I also tried an array of shorts as all the individual values are int16, but this also didn't work. 
Also I am fairly new to Milo and OPC-UA maybe you could explain a little how this works with these parameters that the identifier of the 4 parameter is a referencing a node. (This is at least what I think so)

**Comment by kevinherron** [comment]
This will be a little complicated because you have to actually build and encode a COORD struct to pass as the argument value.

To start, you'll need to configure your client to read the DataType Dictionary so that you can encode/decode structures defined by the server, if you haven't already. It will look something like this: https://github.com/eclipse/milo/blob/17a8fc8d14893e3e4532a9237789363b8cc50891/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/UnifiedAutomationReadCustomDataTypeExample.java#L43-L48

You'll also need a `DataTypeTree` instance. There's a `DataTypeTreeSessionInitializer` that you can add as a session initializer similar to the `DataTypeDictionarySessionInitializer` above and then get it from a session like this:

```
DataTypeTree dataTypeTree = session.getAttribute(DataTypeDictionarySessionInitializer.SESSION_ATTRIBUTE_KEY);
```

Then you'll have to build a generic Struct and encode it:
```
DataTypeTree dataTypeTree = ...; // see above
NodeId dataTypeId = NodeId.parse("ns=3;s=\"OPCUA_methodTaskBring_IDB\".\"UAMethod_InParameters\".\"COORD\"");

Struct struct = Struct.builder("COORD")
    .addMember("L", 0.0)
    .addMember("S", 0.0)
    .addMember("X", 0.0)
    .addMember("Y", 0.0)
    .addMember("Z", 0.0)
    .build();

ExtensionObject xo = ExtensionObject.encodeDefaultBinary(
    client.getDynamicSerializationContext(),
    struct,
    dataTypeTree.getDataType(dataTypeId).getBinaryEncodingId()
);
```

Make sure you use the correct datatypes for the structure member values; I don't know what they actually are and just put doubles in as placeholders.

The finally you can use `xo` as the value in the `Variant` for the last argument.

**Comment by JaSicX** [comment]
@kevinherron thanks for your answer, I have implemented all the necessary bits you mentioned but now I get an null pointer exception on  `dataTypeTree.getDataType(dataTypeId).getBinaryEncodingId());` the getDataType(dataTypeId) returns null. So it dont gets the DataType do I need to add the DataType manually with the DataTypeManger ? Because I thought  
  `client.addSessionInitializer(new DataTypeDictionarySessionInitializer(new GenericBsdParser()));` did that for me.
  
here is the COORD node from UAExpert
![UAexpert3COORD](https://user-images.githubusercontent.com/43895907/133607625-d45ff0c2-5a00-47e2-a7fb-46f5798b6bda.PNG)

and one field 
![UAexpert3L](https://user-images.githubusercontent.com/43895907/133607721-cfccda19-5a37-4bd8-b4f4-4f9359ef19cf.PNG)


the Code is as follows

Code added in the initialization of the client
``` java
        tempOpcUaClient = OpcUaClient.create(opcUaClientconfig);

        tempOpcUaClient.addSessionInitializer(new DataTypeDictionarySessionInitializer(new GenericBsdParser())); 
        tempOpcUaClient.addSessionInitializer(new DataTypeTreeSessionInitializer());

```

Code for constructing the struct 
``` java
 public ExtensionObject buildStructCoords(short[] coords, String node) throws UaRuntimeException, UaException, InterruptedException, ExecutionException {
        OpcUaSession session =  client.getSession().get();
        
        DataTypeTree dataTypeTree = (DataTypeTree) session.getAttribute(DataTypeTreeSessionInitializer.SESSION_ATTRIBUTE_KEY);
        
        NodeId dataTypeId = NodeId.parse(node);
        
        Struct struct = Struct.builder("COORD")
                .addMember("L", coords[0])
                .addMember("S", coords[1])
                .addMember("X", coords[2])
                .addMember("Y", coords[3])
                .addMember("Z", coords[4])
                .build();
        
        ExtensionObject structEncoded = ExtensionObject.encodeDefaultBinary(client.getDynamicSerializationContext()
                , struct, 
                dataTypeTree.getDataType(dataTypeId).getBinaryEncodingId());
        return structEncoded;
                        
    }

```

**Reply by kevinherron** [reply]
I’d start by making sure this server supports the DataType Dictionary and that it’s enabled - I know in e.g. the S7-1500 PLCs it’s something that is optionally enabled in later firmware versions. I can’t tell if that’s what your’e connected to or if it’s some other Siemens product.

**Reply by kevinherron** [reply]
See https://support.industry.siemens.com/cs/document/109765509/how-do-you-best-configure-an-opc-ua-client-to-communicate-efficiently-with-the-opc-ua-server-of-a-simatic-s7-1500-cpu-and-with-the-best-performance-possible-?dti=0&lc=en-WW

> The setting for the "Type Dictionary" is in the device configuration of the CPU under "OPC UA > General > Backward compatible data type definitions according to OPC UA specification <= V1.03".

**Comment by mynickistak** [comment]
@kevinherron Thanks!
> The setting for the "Type Dictionary" [...]

was turned off indeed. It's enabled now but we still have the same problem.

We receive a DataTypeTree as follows:
`DataTypeTree dataTypeTree = (DataTypeTree) session.getAttribute(DataTypeTreeSessionInitializer.SESSION_ATTRIBUTE_KEY);
`
but 
>  I get an null pointer exception on dataTypeTree.getDataType(dataTypeId).getBinaryEncodingId()); the getDataType(dataTypeId) returns null.

Our current idea is to construct or add the DataType manually. Or do you have some suggestions? Could it still be an PLC issue?


EDIT:
@Kevin Herron We discovered this GitHub question:
[how-to-get-the-datatype-of-an-uanode](https://stackoverflow.com/questions/64985967/how-to-get-the-datatype-of-an-uanode)

Solution from this guy named Kevin Herron ;) worked fine, thank you!

```java
node = "ns=3;s=\"OPCUA_methodTaskBring_IDB\".\"UAMethod_InParameters\".\"COORD\"";
NodeId dataTypeId = NodeId.parse(node);
UaVariableNode coordsNode = client.getAddressSpace().getVariableNode(dataTypeId);
NodeId dataType = coordsNode.getDataType();
DataType dT = dataTypeTree.getDataType(dataType);

ExtensionObject structEncoded = ExtensionObject.encodeDefaultBinary(client.getDynamicSerializationContext(), struct,
                dT.getBinaryEncodingId());
```

**Reply by kevinherron** [reply]
So everything is working now? Was the DataType NodeId incorrect and that's why it was returning null?

edit: yeah, that's what it was. I thought the original NodeId you posted was pointing at the DataType Node, but it was just the id of the Variable Node. I should have noticed that.

**Reply by mynickistak** [reply]
Yes everything is working.
